### ISSUE TITLE
described enums underlying mode

### ISSUE THREAD

**Issue by KostinPavel** [boostorg/json#866]
### Version of Boost 

1.81

### Description

Extraction from "underlying" value does not work. Also not covered are "signed underlying" and "unsigned underlying".

### Steps necessary to reproduce the problem

```
#include <boost/describe.hpp>
#include <boost/json.hpp>
#include <boost/json/src.hpp>

namespace my_namespace
{
enum class enum_class_use_underlying_type : uint8_t
{
  var0 = 0,
  var1 = 1
};
BOOST_DESCRIBE_ENUM(enum_class_use_underlying_type)
}  // namespace my_namespace

#include <iostream>

int main()
{
  try {
    const my_namespace::enum_class_use_underlying_type
        enum_class_use_underlying_value {
            my_namespace::enum_class_use_underlying_type::var0};

    const boost::json::value json_value =
        boost::json::value_from(enum_class_use_underlying_value);

    std::cout << "enum_class_use_underlying_value: " << json_value
              << std::endl;  // OK

    const auto enum_class_use_underlying_value2 =
        boost::json::value_to<my_namespace::enum_class_use_underlying_type>(
            json_value);  // Bug

    return 0;
  } catch (const std::exception& e) {
    std::cerr << e.what() << std::endl;
  }
  return 1;
}
```
Output:
```
enum_class_use_underlying_value: 0
value is not a string [boost.json:30 at /usr/include/boost/json/detail/value_to.hpp:558:9 in function 'boost::json::result<T> boost::json::detail::value_to_impl(boost::json::try_value_to_tag<T>, const boost::json::value&, described_enum_conversion_tag)']
```

Because in "boost/json/detail/value_from.hpp" function value_from_helper correct use condition `if( name )`:
```
// described enums
template<class T>
void
value_from_helper(
    value& jv,
    T from,
    described_enum_conversion_tag)
{
    (void)jv;
    (void)from;
#ifdef BOOST_DESCRIBE_CXX14
    char const* const name = describe::enum_to_string(from, nullptr);
    if( name )
    {
        string& str = jv.emplace_string();
        str.assign(name);
    }
    else
    {
        using Integer = typename std::underlying_type< remove_cvref<T> >::type;
        jv = static_cast<Integer>(from);
    }
#endif
}
```
but in "boost/json/detail/value_to.hpp" function value_to_impl not using condition `if( name )`:
```
// described enums
template<class T>
result<T>
value_to_impl(
    try_value_to_tag<T>,
    value const& jv,
    described_enum_conversion_tag)
{
    T val = {};
    (void)jv;
#ifdef BOOST_DESCRIBE_CXX14
    error_code ec;

    auto str = jv.if_string();
    if( !str )
    {
        BOOST_JSON_FAIL(ec, error::not_string);
        return {system::in_place_error, ec};
    }

    if( !describe::enum_from_string(str->data(), val) )
    {
        BOOST_JSON_FAIL(ec, error::unknown_name);
        return {system::in_place_error, ec};
    }
#endif

    return {system::in_place_value, val};
}
```

**Comment by grisumbras** [comment]
First, what is "'underlying' value"?

Second, at first I thought that I forgot to document that conversion happens from **described** enumerators. But I did document it. So, everything works as documented. Did you forget to describe the enumerators? What are you describing the enum for, if not for enumerators?

What is the behaviour you expected?

**Comment by KostinPavel** [comment]
In declare: "enum class enum_class_use_underlying_type : uint8_t"
uint8_t - is underlying

**Comment by grisumbras** [comment]
That's a type, not a value. What's the behaviour you expected?

**Comment by KostinPavel** [comment]
Non-descriptive names should be serialized and deserialized as numbers.

**Comment by grisumbras** [comment]
Ok, so currently the library behaves as documented. And, I guess you ask for a different behaviour?

I'll explain the rationale for our current behaviour. Library-provided conversions are generic and thus try to be the most logical for the types they support. It's logical that for described enums users want to convert to and from strings denoting enumerators.

Converting _from_ a random number to some generic enum is problemtaic, because you cannot programmatically determine the allowed range of values for the enum (it's not always the full domain of the underlying type). Thus we decided to _not allow_ such conversion.

Converting from an enum to its underlying type is never a problem, and we did not want to throw from `value_from`, so we decided to _allow_ such a conversion.

So, now that you know the rationale, what's your rationale for describing an enum, but not its enumerators? If you want
to just represent enum as a number, you can always provide `value_to/from` overloads

```c++
namespace my_namespace
{

void value_from( ::boost::json::value_from_tag, value& jv, enum_class_use_underlying_type e )
{
    using UnderlyingType = std::underlying_type< enum_class_use_underlying_type >::type;
    jv = static_cast< UnderlyingType >(e);
}

enum_class_use_underlying_type
value_from( ::boost::json::value_to_tag<enum_class_use_underlying_type>, value const& jv )
{
    using UnderlyingType = std::underlying_type< enum_class_use_underlying_type >::type;
    auto const n = jv.to_number< UnderlyingType >();
    return static_cast<enum_class_use_underlying_type>(n);
}

}
```

**Comment by KostinPavel** [comment]
All right. I have to use a solution similar to yours. Yes, I had to enter ranges and/or lists of valid values for "enum". But still there should be a built-in way for the non-string representation of "enum" to ensure compatibility with previous solutions.

**Comment by grisumbras** [comment]
What previous solutions? We never provided a generic conversion for described (or any other kind of) enums with a different behaviour.

**Comment by KostinPavel** [comment]
Suppose we are expecting a string, but there is a number. It is necessary to at least try to return a value, even without checking the validity of the value. Top-level code can additionally check if the value is correct.
While in "describe" I did not see the built-in possibility of such a solution.

**Comment by KostinPavel** [comment]
> What previous solutions? We never provided a generic conversion for described (or any other kind of) enums with a different behaviour.

For example, the client and server parts may use different libraries.

**Comment by grisumbras** [comment]
Ok, so this _does not_ sound like a generic conversion. This is a specific conversion for specific protocol. So, you will have to implement it yourself. Since you already have unearthed implementation for our described enum conversions, you can use it as the basis for your implementation.

**Comment by KostinPavel** [comment]
I have my own solution, and I have been successfully using it for a long time. But then I think that people would not be confused, it is necessary to remove the direct conversion to a number.
Then they will immediately understand that they need a specific solution for enums.

**Comment by grisumbras** [comment]
I shared your concern when we implemented this. But currently we don't throw from our generic conversions to `value` (unless it's a memory allocation error), so we do not have a way to communicate the error to the user. Maybe I should add a note in the docs to explain the situation with described enums?

**Comment by KostinPavel** [comment]
For example, you can do something like this:
```
// described enums
template<class T>
result<T>
value_to_impl(
    try_value_to_tag<T>,
    value const& jv,
    described_enum_conversion_tag)
{
    T val = {};
    (void)jv;
#ifdef BOOST_DESCRIBE_CXX14
    error_code ec;

    auto str = jv.if_string();
    auto num = jv.if_int64();
    auto unum = jv.if_uint64();
    if(str)
    {
      if( !describe::enum_from_string(str->data(), val) )
      {
          BOOST_JSON_FAIL(ec, error::unknown_name);
          return {system::in_place_error, ec};
      }
    }
    else if( num )
    {
       char const* const name = describe::enum_to_string(*num, nullptr);
       if( name )
       {
         BOOST_JSON_FAIL(ec, error::not_string);
         return {system::in_place_error, ec};
       }
       val =  static_cast<T>(*num);
    }
    else if( unum )
    {
       char const* const name = describe::enum_to_string(*unum, nullptr);
       if( name )
       {
         BOOST_JSON_FAIL(ec, error::not_string);
         return {system::in_place_error, ec};
       }
       val = static_cast<T>(*unum);
    }
```

    BOOST_JSON_FAIL(ec, error::exhausted_variants);
    return {system::in_place_error, ec};

#endif

    return {system::in_place_value, val};
}

**Comment by KostinPavel** [comment]
Worse, I think it's not being done. Those who need to check the values for validity will do this separately.

**Comment by KostinPavel** [comment]
Alternatively, you can add a #define that will serialize and deserialize as a number. In the #define description, you can specify features...

**Comment by grisumbras** [comment]
This
```
val =  static_cast<T>(*num);
```
is undefined behaviour for many enums. And we don't have a way to check if it is or isn't.

**Comment by KostinPavel** [comment]
Yes, I know. But it comes from compilers and language. You use pointers - although they can also fail. This feature is not easy to take into account - it should not be an obstacle to the operation of the code.

**Comment by KostinPavel** [comment]
Someday the check will be built into the language, then everything will work as it should. And now it's stupid to stop the development of the code because of this reason.

**Comment by KostinPavel** [comment]
Those who need more guarantees now can take care of it without your checking.

**Comment by KostinPavel** [comment]
Checks can and do for pointers as well. A value missing from the enum will not crash the program. This value will be treated as other values. Programmers in code may or may not handle other values. It's not bad - it's just a special case.

**Comment by KostinPavel** [comment]
the only thing to worry about is the size of the type. Might need to add
std::numeric_limits<std::underlying_type_t<T>>::min()
std::numeric_limits<std::underlying_type_t<T>>::max()

**Comment by KostinPavel** [comment]
For example like this:
```
// described enums
template<class T>
result<T>
value_to_impl(
    try_value_to_tag<T>,
    value const& jv,
    described_enum_conversion_tag)
{
    T val = {};
    (void)jv;
#ifdef BOOST_DESCRIBE_CXX14
    error_code ec;

    auto str = jv.if_string();
    auto num = jv.if_int64();
    auto unum = jv.if_uint64();
    if(str)
    {
      if( !describe::enum_from_string(str->data(), val) )
      {
          BOOST_JSON_FAIL(ec, error::unknown_name);
          return {system::in_place_error, ec};
      }
    }
    else if( num )
    {
       char const* const name = describe::enum_to_string(*num, nullptr);
       if( name )
       {
         BOOST_JSON_FAIL(ec, error::not_string);
         return {system::in_place_error, ec};
       }
       if(*num < std::numeric_limits<std::underlying_type_t<T>>::max() || *num > std::numeric_limits<std::underlying_type_t<T>>::max())
       {
         BOOST_JSON_FAIL(ec, error::size_mismatch);
         return {system::in_place_error, ec};
       }       
       val =  static_cast<T>(*num);
    }
    else if( unum )
    {
       char const* const name = describe::enum_to_string(*num, nullptr);
       if( name )
       {
         BOOST_JSON_FAIL(ec, error::not_string);
         return {system::in_place_error, ec};
       }
       if(*unum < std::numeric_limits<std::underlying_type_t<T>>::max() || *unum > std::numeric_limits<std::underlying_type_t<T>>::max())
       {
         BOOST_JSON_FAIL(ec, error::size_mismatch);
         return {system::in_place_error, ec};
       }
       val = static_cast<T>(*unum);
    }

    BOOST_JSON_FAIL(ec, error::exhausted_variants);
    return {system::in_place_error, ec};

#endif

    return {system::in_place_value, val};
}
```

**Comment by grisumbras** [comment]
There are actual compiler optimizations that take advantage of the allowed enum range. Your program _may_ crash. It _may_ behave erratically. That's what "undefined behaviour" means.

**Comment by KostinPavel** [comment]
The above example addresses this issue.

**Comment by grisumbras** [comment]
There's no point in explaining to me how to amend the current implementation to get this or that behaviour. I've wrote those functions, I know how to change them. But I'm telling you, we're not adding behaviour that silently enables UB.

**Comment by grisumbras** [comment]
No, it does not

**Comment by grisumbras** [comment]
```
enum E {a=1, b=2};
int n = 4;
E e = static_cast<E>(n); // UB
```

**Comment by KostinPavel** [comment]
Explain to me, please, why?
What am I not considering?

**Comment by KostinPavel** [comment]
> ```
> enum E {a=1, b=2};
> int n = 4;
> E e = static_cast<E>(n); // UB
> ```

Plice add std::numeric_limits<std::underlying_type_t>::min() std::numeric_limits<std::underlying_type_t>::max() testto this code.

**Comment by KostinPavel** [comment]
```
if(n <std::numeric_limit<std::underlying_type_t<E>::min() || n >std::numeric_limit<std::underlying_type_t<E>::max())
{
throw ....
}
E e = static_cast<E>(n); // correct
  switch (e)
  {
  case E::a:
    /* code */
    break;
  case E::b:
    /* code */
    break;
  
  default:
    /* over values */
    break;
  }
```

**Comment by grisumbras** [comment]
For an enum without a fixed unerlying type the range of its values is the range of values that is possible to represent with the smallest bitfiled that can represent every enum's enumerator. See here: https://eel.is/c++draft/enum#dcl.enum-8

The allowed range for `E`, in this case is [1; 3]. 4 is outside the allowed range. But it is not outside the allowed range of the underlying type, which is not specified, but on most implementations will be at least [0; 256] or [-128; 127].

Look at this example: https://godbolt.org/z/Toj6WTbda
Notice how the function is compiled to never return "unnkown", even though the underlying type clearly supprorts values larger than 3.

**Comment by KostinPavel** [comment]
try it in [online](https://godbolt.org/)
```
#include <limits>
#include <type_traits>
#include <stdexcept>
#include <iostream>


enum E {a=1, b};

static_assert( std::numeric_limits< std::underlying_type_t<E> >::max() > 4 );

char const* message(E e) {
    switch( e ) {
    case a: return "a";
    case b: return "b";
    case E(3): return "3";
    default: return "unknown";
    }
}

int main()
{
    try
    {
        int n = 4;
        if(n < std::numeric_limits<std::underlying_type_t<E> >::min() || n > std::numeric_limits<std::underlying_type_t<E> >::max())
        {
            throw std::runtime_error("overflow");
        }
        E e = static_cast<E>(n);    
    } catch (const std::exception& e) {
    std::cerr << e.what() << std::endl;
    }
  return 0;
}
```

**Comment by grisumbras** [comment]
What does that snippet examplify? If your check would have made a difference an exception would have been thrown. But it doesn't happen. https://godbolt.org/z/Pa7989Y5P

**Comment by KostinPavel** [comment]
Compare to this:
```
#include <limits>
#include <type_traits>
#include <stdexcept>
#include <iostream>


enum E {a=1, b};

static_assert( std::numeric_limits< std::underlying_type_t<E> >::max() > 4 );

int main()
{
    E e = a;
    try
    {
        std::cout << std::numeric_limits<std::underlying_type_t<E> >::min() << std::endl;
        std::cout << std::numeric_limits<std::underlying_type_t<E> >::max() << std::endl;
        uint64_t n = 4294967295 + 10;
        if(n < std::numeric_limits<std::underlying_type_t<E> >::min() || n > std::numeric_limits<std::underlying_type_t<E> >::max())
        {
            throw std::runtime_error("overflow");
        }
        e = static_cast<E>(n);    
    } catch (const std::exception& e) {
        std::cerr << e.what() << std::endl;
    }
    return 0;
}
```

**Comment by KostinPavel** [comment]
This demonstrates that UB is being processed.

**Comment by KostinPavel** [comment]
Any value outside the definition will either default case or rise exception an overflow. Both situations are expected.

**Comment by grisumbras** [comment]
No, the value 4 does not result in a thrown exception. You seem to ignore what I'm saying you. Are you trying to say that the value 4 is allowed for enum E?

**Comment by KostinPavel** [comment]
Why should?
enum {
a=1,
b=5
};

**Comment by KostinPavel** [comment]
"4" - it's just another value.

**Comment by grisumbras** [comment]
No, 4 is outside of the range of [0;3] which is the range of enum E {a=1, b};

**Comment by KostinPavel** [comment]
Imagine that it is simply not described. Perhaps up to a certain point in time.

**Comment by KostinPavel** [comment]
You have a slightly distorted idea of enumerations. How then do you explain the missing definitions to yourself?
**enum {
a=1,
b=5
};**

**Comment by KostinPavel** [comment]
Treat them as numbers, some of which have a name.

**Comment by grisumbras** [comment]
I don't have a distorted idea of unumerations. I've linked to a clause in the C++ standard where it is explained what a valid range of allowed values for an enumeration is. For enum {a=1, b = 5} the valid range is [0; 7] (in other words, everything that can be represented by 3 bits, which is how many bits required for the number 5).

**Comment by grisumbras** [comment]
Here, a crash. https://godbolt.org/z/s8sjqxqa6

**Comment by KostinPavel** [comment]
But the standard does not say anything about what to do with other values. Including missed ones.
And if the compiler allocated 3 bits for enumeration, then std::numeric_limits<std::underlying_type_t<E> >::min() and std::numeric_limits<std::underlying_type_t<E> >::max() would beat the same by others.

**Comment by KostinPavel** [comment]
Why do you offer me a static_cast. It is obvious that it does not work while at a compilation stage.

**Comment by grisumbras** [comment]
What are you talking about? It compiles and crashes.

**Comment by KostinPavel** [comment]
> What are you talking about? It compiles and crashes at runtime.
This is a programmer's mistake. He didn't check the range.

**Comment by KostinPavel** [comment]
The proposed solution has a check.

**Comment by KostinPavel** [comment]
> Here, a crash. https://godbolt.org/z/s8sjqxqa6

The library will not protect against such an error.

**Comment by KostinPavel** [comment]
Fine. I tried to justify all my thoughts. You decide. The developer is you.
My solution uses the described approach.
Whether or not this approach is available to others is up to you.

**Comment by KostinPavel** [comment]
The only other way I can help is to make a branch with a pull request.

**Comment by grisumbras** [comment]
There's no way to check if that conversion will be UB or not. If you know how to do it, please change the example I provided (the one with a crashing program) so that it does not crash. Then we can discuss whether your change is possible to integrate into the library. Things you've suggested so far _would not_ have solved the problem, because they do not address the problem in any way.

**Comment by grisumbras** [comment]
I repeat, I can't check the range. Unless I'm missing something. Please, show me in CE how to do that, by making the program to not crash.

**Comment by KostinPavel** [comment]
#include <limits>
#include <cstdio>
#include <stdexcept>

enum E {a=1, b};

char const* message(E e) {
    switch( e ) {
    case a: return "a";
    case b: return "b";
    //case E(0): return "0"; // this programmer error use sanitazer to detect this
    //case E(3): return "3"; // this programmer error use sanitazer to detect this
    default: return "something else";
    }
}

int main(int argc, char**)
{    
    try
    {
    argc += 3;     
    if(argc < std::numeric_limits<std::underlying_type_t<E> >::min() || argc > std::numeric_limits<std::underlying_type_t<E> >::max())
    {
        throw std::runtime_error("overflow");
    }
    E e = static_cast<E>(argc);
    std::printf(message(e));
    } catch (const std::exception& e) {
        std::printf(e.what());        
    }
    return 0;
}

**Comment by KostinPavel** [comment]
In this library, there is no need to iterate over the values in switch!

**Comment by grisumbras** [comment]
You did not catch the the value being outside of enum's range. You just changed the code so that UB would not manifest. So, you don't know how to check the range either. Of course, it's because there's no way to do that.

Just to reiterate, I _will not_ add silent undefind behaviour to the library. Users are free to write their own conversions, which can have as much of UB as they want. They also may solve the issue by only using enums with fixed underlying types. But as generic conversions have to be generic, there's no acceptable solution on our end.

**Comment by KostinPavel** [comment]
I think you are wrong.
In the proposed solution, UB also remain on the side of the developer. But they do not need to finish your decision with their crutches.

**Comment by KostinPavel** [comment]
The code remains transparent within the capabilities of the language and the compiler.

**Comment by KostinPavel** [comment]
Do you think someone expects that it will be converted to a number, but not back?

**Comment by grisumbras** [comment]
It's defined by definition. Because it is documented. I will consider adding a note to make this more explicit.

**Comment by KostinPavel** [comment]
This is code from the library's user space!
```
char const* message(E e) {
    switch( e ) {
    case a: return "a";
    case b: return "b";
    //case E(0): return "0"; // this programmer error use sanitazer to detect this
    //case E(3): return "3"; // this programmer error use sanitazer to detect this
    default: return "something else";
    }
}
'''

**Comment by KostinPavel** [comment]
"-fstrict-enums" also set by the library user

**Comment by KostinPavel** [comment]
example of a warning to a library user:
C:\msys64\mingw64\bin\g++.exe main.cpp -o d:\VSCode\CPP\enum_test\output\main.exe -Wall -Wextra -g3 

main.cpp: In function 'const char* message(E)':
main.cpp:10:5: warning: case value '0' not in enumerated type 'E' [-Wswitch]
   10 |     case E(0): return "0";
      |     ^~~~
main.cpp:11:5: warning: case value '3' not in enumerated type 'E' [-Wswitch]
   11 |     case E(3): return "3";
      |     ^~~~

**Comment by KostinPavel** [comment]
> > Here, a crash. https://godbolt.org/z/s8sjqxqa6
> 
> The library will not protect against such an error.
Maybe a bug in https://godbolt.org/
Because it worked for me as expected.
C:\msys64\mingw64\bin\g++.exe main.cpp -o d:\VSCode\CPP\enum_test\output\main.exe -std=c++20 -O3 -fstrict-enums 
PS D:\VSCode\CPP\enum_test> cd 'd:\VSCode\CPP\enum_test\output'
PS D:\VSCode\CPP\enum_test\output> & .\'main.exe'
3
PS D:\VSCode\CPP\enum_test\output>

**Comment by KostinPavel** [comment]
Crashes because argc is not set.

```
int main(int argc, char**)
{
    argc = 0;
    E e = static_cast<E>(argc + 3);
    std::printf(message(e));
    return 0;
}
```

**Comment by KostinPavel** [comment]
Please let us know if it worked for you.

**Comment by KostinPavel** [comment]
I also tried it:
```
#include <limits>
#include <cstdio>


enum E {a=1, b};

char const* message(E e) {
    switch( e ) {
    case a: return "a";
    case b: return "b";
    case E(0): return "0";
    case E(3): return "3";
    default: return "something else";
    }
}

int main(int argc, char**)
{
    argc = std::numeric_limits<int>::max();
    std::printf("%ull\n",argc);
    std::printf("%ull\n",argc+3);
    E e = static_cast<E>(argc + 3);
    std::printf(message(e));
    return 0;
}
```
It works too

**Comment by KostinPavel** [comment]
I don't think you'll find a case that doesn't work.

**Comment by pdimov** [comment]
> This
> 
> ```
> val =  static_cast<T>(*num);
> ```
> 
> is undefined behaviour for many enums.

In this specific case not, because the value has been checked to have a corresponding enumerator (via enum_to_string).

**Comment by pdimov** [comment]
> ```c++
> enum E {a=1, b};
> 
> char const* message(E e) {
>     switch( e ) {
>     case a: return "a";
>     case b: return "b";
>     case E(0): return "0";
>     case E(3): return "3";
>     default: return "something else";
>     }
> }
> ```

This is OK because 3 is in the range of the enum (which is 0..3 in this case because it has to contain the values 1 and 2). But E(4) would be UB.

**Comment by pdimov** [comment]
The rule is, if the value can't fit in the allowed range of the enum, behavior is undefined. The allowed range is: if the enum has a fixed underlying type, the range is that of the type; otherwise, the range is the range of the hypothetical integral type with the smallest number of bits that can fit all the enumerators. So

```
enum E1 { a = 1, b = 5 }; // range is 0..7
enum E2: int {}; // range is that of int
enum class E3: int {}; // range is that of int
enum class E4 {}; // range is that of int, because enum class always has a fixed underlying type
```

So in principle, we can check is_scoped_enum (which is implementable as in https://github.com/boostorg/endian/blob/develop/include/boost/endian/detail/is_scoped_enum.hpp), and there's no UB there for any `underlying_type` value.

But there's no reliable way to detect is_fixed_enum (E2 above.) I have a trait that kind of does (https://github.com/boostorg/endian/blob/feature/is-fixed-enum/include/boost/endian/detail/is_fixed_enum.hpp) but it requires C++17 and doesn't quite work correctly on all compilers.

And there's absolutely no way to obtain the range of E1 (except if it's correctly described, in which case we can compute it from the enumerator descriptors).

**Comment by KostinPavel** [comment]
Perhaps you need to make implementations for each of the four options.
```c++
enum E1 { a = 1, b = 5 }; // range is 0..7 //most compilers will choose a chunk of memory that is a multiple of a byte
enum E2: int {}; // range is that of int
enum class E3: int {}; // range is that of int // does not require discussion, since the solution is proposed
enum class E4 {}; // range is that of int, because enum class always has a fixed underlying type //does not require discussion, since the solution is proposed
```

**Comment by KostinPavel** [comment]
It's a good idea to compare values from string descriptions, but it's not clear how to convert to a number then?

**Comment by pdimov** [comment]
> ```c++
> // ??? most compilers will choose a chunk of memory that is a multiple of a byte for this type
> ```

It's still undefined behavior. Bad things will happen if not today, tomorrow.

In fairness, -fsanitize=undefined doesn't diagnose this for both GCC and Clang, and it seems to be unspecified before C++17, not undefined.

**Comment by KostinPavel** [comment]
to detect the numeric mode, you can use:
```c++
//default
template<typename T>
constexpr bool need_numeric_mode = false;

//specialization:
template<my_enum_type>
constexpr bool need_numeric_mode = true;
```

**Comment by KostinPavel** [comment]
The structure of the code is something like this:
```c++
// described enums
template<class T>
result<T>
value_to_impl(
    try_value_to_tag<T>,
    value const& jv,
    described_enum_conversion_tag)
{
    T val = {};
    (void)jv;
#ifdef BOOST_DESCRIBE_CXX14
    error_code ec;

    auto str = jv.if_string();
   
    if constexpr (need_numeric_mode<T>)
    {
        auto num = jv.if_int64();
        auto unum = jv.if_uint64();
        if( num )
        {
            char const* const name = describe::enum_to_string(*num, nullptr);
            if( !name )
            {
                BOOST_JSON_FAIL(ec, error::value);
                return {system::in_place_error, ec};
            }
            val =  static_cast<T>(*num);
        }
        else if( unum )
        {
            char const* const name = describe::enum_to_string(*unum, nullptr);
            if( !name )
            {
                BOOST_JSON_FAIL(ec, error::value);
                return {system::in_place_error, ec};
            }
            val =  static_cast<T>(*unum);
        }
        else
        {
            BOOST_JSON_FAIL(ec, error::not_number);
            return {system::in_place_error, ec};
        }
    }
    else
    {
        if(!str)
        {
            BOOST_JSON_FAIL(ec, error::not_string);
            return {system::in_place_error, ec};
        }
        if( !describe::enum_from_string(str->data(), val) )
        {
            BOOST_JSON_FAIL(ec, error::unknown_name);
            return {system::in_place_error, ec};
        }
    }

#endif

    return {system::in_place_value, val};
}
```

**Comment by KostinPavel** [comment]
Something similar is needed when converting to a number.
